client/x/evmstaking/keeper/abci.go

```diff
@@ -132,5 +127,47 @@ func (k *Keeper) EndBlock(ctx context.Context) (abci.ValidatorUpdates, error) {
 		return nil, err
 	}
 
+	// Update validator set
+	var valUpdates []abci.ValidatorUpdate
+
+	isNextEpoch, err := k.IsNextEpoch(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "check if the next epoch starts")
+	}
+
+	//nolint:nestif // readability
+	if isNextEpoch {
+		// get queued messages
+		queuedMsgs, err := k.DequeueAllMsgs(ctx)
```

Comment:
I'm thinking to improve the performance, we can combine the DequeueAllMsgs and ProcessMsg. Because we don't actually need to dequeue and store in a separate data structure.

We just need to iterate through the queue and directly process each message. In the end, you initialize a new message queue anyway so no need to dequeue.

Reply:
Got it. valid point.
You mean, at the end of epoch, iterate all queued message with direct execution (not dequeue) and after that initialize it, right?

Reply:
Could you check this [99c399e](https://github.com/piplabs/story/pull/96/commits/99c399e230849d3e27015e4532b28faef53f48b0)?
I merge two funcs and pushed as a separate commit.

Reply:
Yes! This looks good to me.